database/src/bin/postgres-to-sqlite.rs

```diff
@@ -0,0 +1,612 @@
+//! Program to export a Postgres database to SQLite.
+//!
+//! This is intended to be used on an empty SQLite database with no ongoing
+//! transactions, and will likely fail if used on a populated database.
+
+use chrono::{DateTime, Utc};
+use clap::value_t;
+use database::pool::{postgres, sqlite, ConnectionManager};
+use futures::StreamExt;
+use rusqlite::params;
+use std::time::Instant;
+
+const ARTIFACT_WHERE: &str = "where artifact.date > (CURRENT_TIMESTAMP - interval '{} week')";
+
+const ARTIFACT_JOIN_AND_WHERE: &str = "join artifact on artifact.id = aid \
+    where artifact.date > (CURRENT_TIMESTAMP - interval '{} week')";
+
+fn with_filter_clause_maybe(
+    mut statement: String,
+    filter_clause: &str,
+    since_weeks_ago: Option<u32>,
+) -> String {
+    if let Some(weeks) = since_weeks_ago {
+        statement.push(' ');
+        statement.push_str(&filter_clause.replace("{}", &weeks.to_string()));
+    }
+
+    statement
+}
+
+trait Table {
+    /// Table name.
+    fn name(&self) -> &'static str;
+
+    // Statement to select rows from Postgres table.
+    //
+    // If `since_weeks_ago` is `Some(n)`, the select statement excludes any rows
+    // associated with artifacts whose date value precedes `n` weeks ago.
+    // Otherwise, it includes all rows.
+    fn postgres_select_statement(&self, since_weeks_ago: Option<u32>) -> String;
+
+    /// Parameterized statement to insert row into SQLite table.
+    fn sqlite_insert_statement(&self) -> &'static str;
+
+    /// Extracts attribute values from Postgres row and inserts them into SQLite
+    /// using the given statement and transaction. The statement should be one
+    /// prepared from the output of `sqlite_insert_statement`.
+    fn sqlite_execute_insert(&self, statement: &mut rusqlite::Statement, row: tokio_postgres::Row);
+}
+
+struct Artifact;
+
+impl Table for Artifact {
+    fn name(&self) -> &'static str {
+        "artifact"
+    }
+
+    fn postgres_select_statement(&self, since_weeks_ago: Option<u32>) -> String {
+        let s = "select id, name, date, type from ".to_string() + self.name();
+        with_filter_clause_maybe(s, ARTIFACT_WHERE, since_weeks_ago)
+    }
+
+    fn sqlite_insert_statement(&self) -> &'static str {
+        "insert into artifact (id, name, date, type) VALUES (?, ?, ?, ?)"
```

Comment:
So I guess this'll error out if there's an ID collision when exporting into a pre-existing db...

This seems OK, there's not really an amazing way to integrate the two name spaces without some pretty intensive mapping work (probably not worth it, we don't really need to migrate DBs into prefilled ones, I suspect, at least not in a completely automated fashion)

Reply:
Yeah, definitely made the design decision to not attempt to handle this. For things with numeric id PKs, it's probably not too hard to handle. But for things where identity is based on content -- e.g. the `benchmark` PK is `name` -- it's not obvious what to do if the "same" row in each DB disagrees on the value of a non-key column. At least not in general.